pkg/destroy/nutanix/nutanix.go

```diff
@@ -0,0 +1,229 @@
+	ec := createExpectedCategory(o.infraID)
+	matchedVirtualMachineList := make([]*nutanixclientv3.VMIntentResource, 0)
+	allVMs, err := o.v3Client.V3.ListAllVM(emptyFilter)
+	if err != nil {
+		return err
+	}
+
+	for _, v := range allVMs.Entities {
+		if hasCategoryAssigned(v.Metadata, ec) {
+			matchedVirtualMachineList = append(matchedVirtualMachineList, v)
+		}
+	}
+
+	if len(matchedVirtualMachineList) == 0 {
+		o.logger.Infof("no VMs found that require deletion for cluster %s", o.clusterID)
+	} else {
+		logToBeDeletedVMs(matchedVirtualMachineList, o.logger)
+		err := deleteVMs(o.v3Client.V3, matchedVirtualMachineList, o.logger)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanupImages(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	allImages, err := o.v3Client.V3.ListAllImage(emptyFilter)
+	if err != nil {
+		return err
+	}
+
+	allErrs := field.ErrorList{}
+	for _, image := range allImages.Entities {
+		if hasCategoryAssigned(image.Metadata, ec) {
+			imageName := *image.Spec.Name
+			imageUUID := *image.Metadata.UUID
+			o.logger.Infof("deleting image %s with UUID %s", imageName, imageUUID)
+			response, err := o.v3Client.V3.DeleteImage(imageUUID)
+			if err != nil {
+				o.logger.Errorf("failed to delete image %s: %q", imageUUID, err)
+				allErrs = append(allErrs, &field.Error{Type: field.ErrorTypeInternal, Detail: err.Error()})
+				continue
+			}
+
+			if err = nutanixtypes.WaitForTask(o.v3Client.V3, response.Status.ExecutionContext.TaskUUID.(string)); err != nil {
+				o.logger.Errorf("failed to confirm image deletion %s: %q", imageUUID, err)
+				allErrs = append(allErrs, &field.Error{Type: field.ErrorTypeInternal, Detail: err.Error()})
+			}
+		}
+	}
+
+	return allErrs.ToAggregate()
+}
+
+func cleanupCategories(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	_, err := o.v3Client.V3.GetCategoryKey(ec.key)
+	if err != nil {
+		if strings.Contains(err.Error(), "does not exist") {
+			//Already deleted
+			return nil
+		}
+		return err
+	}
+
+	allErrs := field.ErrorList{}
+	for _, val := range ec.values {
+		o.logger.Infof("deleting category value : %s", val)
+		err := o.v3Client.V3.DeleteCategoryValue(ec.key, val)
+		if err != nil {
+			o.logger.Errorf("failed to delete category value %s: %q", val, err)
+			allErrs = append(allErrs, &field.Error{Type: field.ErrorTypeInternal, Detail: err.Error()})
+		}
+	}
+
+	o.logger.Infof("deleting category key : %s", ec.key)
+	err = o.v3Client.V3.DeleteCategoryKey(ec.key)
+	if err != nil {
+		o.logger.Errorf("failed to delete category key %s: %q", ec.key, err)
+		allErrs = append(allErrs, &field.Error{Type: field.ErrorTypeInternal, Detail: err.Error()})
+	}
+
+	return allErrs.ToAggregate()
+}
+
+func createExpectedCategory(infraID string) *expectedCategory {
+	return &expectedCategory{
+		key:    fmt.Sprintf(expectedCategoryKeyFormat, infraID),
+		values: []string{expectedCategoryValueOwned, expectedCategoryValueShared},
```

Comment:
The "shared" value should not be included. We do not want to delete resources with the "shared" value.
```suggestion
		values: []string{expectedCategoryValueOwned},
```

Reply:
We still want to delete category value as that is a prerequisite for deleting the category key.

Reply:
We will look for all values of the category and delete them all instead of deleting one or two.